New release instructions

eteq opened the pull request:
This updates the releasing instructions to reflect the current procedure.  It also deprecates the `zest.releaser` machinery as it is no longer needed.

There's a matching astropy/astropy-tools#14 that implements the machinery to generate a new changelog, which should go in around the same time as this to make sure they are consistent.

cc @astrofrog

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -442,6 +460,46 @@ be fixed soon to a new release milestone.  If the upcoming bug fix release is
+before anything is released.  Hence the following procedure should be added
+to the beginning of the above procedure when this is required.
+
+#. In `astropy-helpers`, create a new release branch "release-<version>".
```
I understand now why this was done - if the subsequent PR into master fails for any reason, we can easily re-edit the history of the release branch, but we can't do that for the maintenance or master branches. So I withdraw my objection to this procedure.

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -442,6 +460,46 @@ be fixed soon to a new release milestone.  If the upcoming bug fix release is
+#. Create the release commit (updating the version info and changelog) in that
+   branch.
+
+#. Push the release branch to github.
```
It would be worth clarifying that this whole process is the same if one is doing an RC for the astropy-helpers - the release branch should be ``release-v<version>rc?`` etc.

eteq replied:
``.. note::`` on this added in the most recent commit

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -442,6 +460,46 @@ be fixed soon to a new release milestone.  If the upcoming bug fix release is
+   updating the helpers to the commit described in the last step (i.e., the
+   head of the astropy-helpers release branch).
+
+#. Wait for Travis to run to ensure that helpers build works with Astropy.
```
Travis and Appveyor - or more generally 'the Continuous Integration services' to be future-proof.

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -442,6 +460,46 @@ be fixed soon to a new release milestone.  If the upcoming bug fix release is
+   head of the astropy-helpers release branch).
+
+#. Wait for Travis to run to ensure that helpers build works with Astropy.
+   If it doesn't, back out the release and fix whatever the problem is before
```
What does back out the release mean?

eteq replied:
yep, that's exactly what I meant - will clarify

astrofrog commented:
One additional script that would be useful (which I could write based on the one I did for the affiliated packages) is one that would make the PRs to the astropy core package to update the astropy-helpers in the master and maintenance branches.

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -442,6 +460,46 @@ be fixed soon to a new release milestone.  If the upcoming bug fix release is
+
+      git checkout master
+      git merge --no-ff release-<version>
+      git tag -s "v<version>" -m "Tagging v<version>"
```
The merge also has to be done into the maintenance branch. Do we tag releases in master or in the maintenance branch?

mhvk commented:
Looked here out of curiosity what it actually involved (no, I'm not volunteering!). One suggestion: maybe use `upstream` instead of `origin` for the actual astropy-hosted branch? At least, elsewhere the suggestion has been to use `origin` for one's own github branch, and `upstream` for the astropy one.

eteq commented:
@astrofrog - definitely agree those scripts would be useful!  Perhaps add an issue for that in astropy-tools that also reminds us to update the instructions when they exist?

(And I think I've addressed all your comments other than the one about "The merge also has to be done into the maintenance branch. Do we tag releases in master or in the maintenance branch?" - will discuss that with you out-of-band because that will probably be more efficient)

@mvhk - good point re `upstream`!  Fixed in the last few commits.

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -442,6 +460,60 @@ be fixed soon to a new release milestone.  If the upcoming bug fix release is
+
+This way the commit of the helpers that is tagged as the release is the same
+commit that the astropy_helpers submodule will be on when the PR to astropy
+testing the release gets merged.
```
Should we add a step to then merge the PR(s) on the core repo?

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -12,231 +12,249 @@ more of the process, if not all.
+   more on this.
 
-        $ git shortlog -s
+#. Make sure that the continuous integration services (E.g., Travis) are passing
```
``E.g.`` -> ``e.g.``

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -12,231 +12,249 @@ more of the process, if not all.
-        $ virtualenv --system-site-packages astropy-release
-        $ source astropy-release/bin/activate
+#. Be sure you're on the branch appropriate for the version you're about to
+   release.  For example, if releasing version 0.2.2 make sure to::
```
Maybe we should update this to a more recent/future example, e.g. 1.2.2?

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -12,231 +12,249 @@ more of the process, if not all.
+   ``-s`` option::
 
-    .. note::
+      $ git tag -s v<version>
```
Since it usually asks for a message, I normally do:

```
git tag -s v<version> -m v<version>
```

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -12,231 +12,249 @@ more of the process, if not all.
-     upload)" enter 'N': zest.releaser does not offer enough control yet over
-     how the register and upload are performed so we will do this manually
-     until the release scripts have been improved.
+      $ git commit -m "Back to development: v<next_version>"
```
Maybe clarify that ``<next_version>`` should be followed by ``.dev`` in the commit message, if necessary? (``Back to development: v1.3.0.dev``)

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -12,231 +12,249 @@ more of the process, if not all.
+      $ conda create -n astropy_release_test_v<verson> numpy
+      $ source activate astropy_release_test_v<verson>
+      $ cd dist
+      $ mkdir test
```
I think this line can be removed?

eteq replied:
good point, and actually this can all be done from straight in the source dir to simplify things

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -12,231 +12,249 @@ more of the process, if not all.
-     which distribution to upload::
+   Assuming everything went smoothly with the last step, we are ready to proceed
+   with the release.  If not, you'll have to back up to before starting this
+   procedure and start over when things are fixed.
```
I usually run a ``git clean -dfx`` again here - since we re-generate the source file again after.

astrofrog replied:
I would also add a paragraph about the case where things don't go smoothly - in this case I usually remove the last two commits and then fix issues, and then start the release process again.

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -12,231 +12,249 @@ more of the process, if not all.
 
- 25. Run the ``upload_script.py`` script in ``astropy-website`` to update the actual
-     web site.
+#. In the astropy *master* branch (not just the changelog), be sure to update
```
should 'changelog' be 'maintenance branch'?

astrofrog commented:
One other question - not sure where this should go (if anywhere). In maintenance branches, do we omit the 'back to development' commit, since it isn't really needed?

eteq commented on docs/development/releasing.rst:
```diff
@@ -442,6 +460,82 @@ be fixed soon to a new release milestone.  If the upcoming bug fix release is
+   "tmp-release-v<version>"::
+
+      $ cd /wherever/you/put/astropy/astropy_helpers
+      $ git branch tmp-release-v<version> <maintenance branch name>
```
(Note to myself from an out-of-band conversation with @astrofrog: instead of a local version, just do ``git push <maintenance branch name>:tmp-release-v<version>``)

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -442,6 +460,82 @@ be fixed soon to a new release milestone.  If the upcoming bug fix release is
+      $ git checkout <maintenance branch name>
+      $ git merge --no-ff tmp-release-v<version>
+      $ git tag -s "v<version>" -m "Tagging v<version>"
+      $ git push upstream --tags <maintenance branch name>
```
Do we need to also mention the ``python setup.py sdist register upload``? Or just refer to the instructions for the core package?

astrofrog commented:
Note (while I'm going through the steps): it might be easier to move to a workflow where we have a separate commit for updating the changelog and the setup.py file, since then it will allow the changelog commits to be cherry-picked across branches.

astrofrog commented on docs/development/releasing.rst:
```diff
@@ -12,231 +12,249 @@ more of the process, if not all.
- 8. Install zest.releaser into the virtualenv; use ``--upgrade --force`` to
-    ensure that the latest version is installed in the virtualenv (if you're
-    running a csh variant make sure to run ``rehash`` afterwards too)::
+      $ git commit -m "Preparing release v<version>"
```
So actually it looks like historically we don't include ``v`` in the version in the 'Preparing' and 'Back to development' commits

eteq commented:
OK, @astrofrog, I think I addressed everything you mentioned so far.  You'll want to check out the commits *after* df10cf579254ca2a2e20f490d2c543b6bf4b0768 to see what I've changed from your last look through
.

> One other question - not sure where this should go (if anywhere). In maintenance branches, do we omit the 'back to development' commit, since it isn't really needed?

I think we *don't* want to omit it.  That way when working in maintenance branches, their version is <tobereleased>.dev, so they will be "later" versions from the currently-installed version.  That makes testing installation on maintenance branches much easier (otherwise they'll have the same version as the *previous* release).

> it might be easier to move to a workflow where we have a separate commit for updating the changelog and the setup.py file, since then it will allow the changelog commits to be cherry-picked across branches.

Good idea - d32eaaad7f8eba19f6392ea92234001c5c5be6bc implements that.  Let me know what you think - if you don't like it after all I can always expunge that commit.

> So actually it looks like historically we don't include v in the version in the 'Preparing' and 'Back to development' commits

(for posterity) in a discussion out-of-band with @astrofrog, we concluded that going forward we *would* do that for consistency.

astrofrog commented:
@eteq - thanks, this looks great! Shall we merge once tests pass, and then I can do a PR if there are any fixes I would like to suggest?

eteq commented:
(said this out-of-band, but:) yes, sounds good!
